### PR TITLE
Backup: init backup, restore and migrate api 

### PR DESCRIPTION
**Pull request by Xieql**
**What type of PR is this?**

/kind api-change
/kind design
/kind documentation

**What this PR does / why we need it**:
add backup sync api 
see #374 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
add backup sync api 
```

### CODE REVIEW

File: pkg/apis/backups/v1alpha1/util.go
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/util.go:1]: util is not a good naming, it is to put some functions that developer have no idea where to put it.

Here, may call common_type.go or shared_type.go

File: pkg/apis/backups/v1alpha1/backup_type.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+// Note: partly copied from https://github.com/vmware-tanzu/velero/blob/v1.11.1/pkg/apis/velero/v1/backup_types.go
+// BackupPolicy defines the specification for a backup policy.
+type BackupPolicy struct {
+	// ResourceFilter specifies which resources should be included in the backup.
+	// It acts as a selective criterion to determine which resources are relevant for backup.
+	// If not set, the backup process will consider all resources. This filter helps in optimizing the backup process by excluding unnecessary data.
+	// +optional
+	ResourceFilter *ResourceFilter `json:"resourceFilter,omitempty"`
+
+	// TODO: consider SnapshotVolumes for backup
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/backup_type.go:78]:
```suggestion
	// TODO: support volume snapshot
```

File: pkg/apis/backups/v1alpha1/util.go
Comment (Xieql) [reply pkg/apis/backups/v1alpha1/util.go:1]: ok

File: pkg/apis/backups/v1alpha1/backup_type.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+// Backup is the schema for the Backup's API.
+type Backup struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              BackupSpec   `json:"spec,omitempty"`
+	Status            BackupStatus `json:"status,omitempty"`
+}
+
+type BackupSpec struct {
+	// TODO: consider add Storage setting for backup
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/backup_type.go:42]: remove, duplicate with below

File: pkg/apis/backups/v1alpha1/backup_type.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	// For example, if you have a specific order for pods, such as "pod1, pod2, pod3" with all belonging to the "ns1" namespace,
+	// and a specific order for persistentvolumes, such as "pv4, pv8", you can use the orderedResources field in YAML format as shown below:
+	// ```yaml
+	// orderedResources:
+	//  pods: "ns1/pod1, ns1/pod2, ns1/pod3"
+	//  persistentvolumes: "pv4, pv8"
+	// ```
+	// +optional
+	// +nullable
+	OrderedResources map[string]string `json:"orderedResources,omitempty"`
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/backup_type.go:96]: why do we need order for backup

File: pkg/apis/backups/v1alpha1/migrate_type.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              MigrateSpec   `json:"spec,omitempty"`
+	Status            MigrateStatus `json:"status,omitempty"`
+}
+
+type MigrateSpec struct {
+	// SourceCluster represents the source cluster for migration.
+	// The user needs to ensure that SourceCluster points to only ONE cluster.
+	// Because the current migration only supports migrating from one SourceCluster to one or more TargetCluster.
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/migrate_type.go:41]: + required

File: pkg/apis/backups/v1alpha1/migrate_type.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	SourceCluster *Destination `json:"sourceCluster"`
+
+	// TargetCluster represents the target clusters for migration.
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/migrate_type.go:44]: + required

File: pkg/apis/backups/v1alpha1/migrate_type.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	TargetCluster *Destination `json:"targetCluster"`
+
+	// Policy defines the rules for the migration.
+	// +optional
+	Policy *MigratePolicy `json:"policy,omitempty"`
+}
+
+type MigratePolicy struct {
+	// ResourceFilter specifies the resources to be included in the migration.
+	// If not set, all resources in  will be migrated.
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/migrate_type.go:54]: in what

File: pkg/apis/backups/v1alpha1/restore_type.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+	// +nullable
+	PreserveNodePorts *bool `json:"preserveNodePorts,omitempty"`
+}
+
+type ReserveStatusSpec struct {
+	// IncludedResources specifies the resources to which will restore the status.
+	// If empty, it applies to all resources.
+	// +optional
+	// +nullable
+	IncludedResources []string `json:"includedResources,omitempty"`
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/restore_type.go:86]: add examples

File: pkg/apis/backups/v1alpha1/restore_type.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+	// ExcludedResources specifies the resources to which will not restore the status.
+	// +optional
+	// +nullable
+	ExcludedResources []string `json:"excludedResources,omitempty"`
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/restore_type.go:91]: add examples

File: pkg/apis/backups/v1alpha1/restore_type.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+	// Any source namespaces not included in the map will be restored into
+	// namespaces of the same name.
+	// +optional
+	NamespaceMapping map[string]string `json:"namespaceMapping,omitempty"`
+
+	// ReserveStatus specifies which resources we should restore the status field.
+	// If unset, no status will be restored.
+	// +optional
+	// +nullable
+	ReserveStatus *ReserveStatusSpec `json:"restoreStatus,omitempty"`
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/restore_type.go:72]: preserveStatus

File: pkg/apis/backups/v1alpha1/restore_type.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+	// PreserveNodePorts specifies whether to restore old nodePorts from backup.
+	// If not specified, default to false.
+	// +optional
+	// +nullable
+	PreserveNodePorts *bool `json:"preserveNodePorts,omitempty"`
+}
+
+type ReserveStatusSpec struct {
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/restore_type.go:81]: PreserveStatusConfig

File: pkg/apis/backups/v1alpha1/util.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+	// ClusterSelector specifies the selectors to select the clusters within the fleet.
+	// If unspecified, all clusters in the fleet will be selected.
+	// The field will only take effect when Fleet is set.
+	// +optional
+	ClusterSelector *ClusterSelector `json:"clusterSelector,omitempty"`
+
+	// Clusters determine a set of clusters as destination clusters.
+	// The field will only take effect when Fleet is not set.
+	// +optional
+	Clusters []*corev1.ObjectReference `json:"clusters,omitempty"`
```
Comment (hzxuzhonghu) [inline pkg/apis/backups/v1alpha1/util.go:41]: so this means clusters will have lower priority

File: pkg/apis/backups/v1alpha1/backup_type.go (hunk 3 above)
Comment (Xieql) [reply pkg/apis/backups/v1alpha1/backup_type.go:96]: The resource backup order feature provided by Velero is primarily designed to address scenarios like master-slave database configurations. In such setups, there's an inherent delay in data synchronization or state propagation. To ensure data consistency and coherence after backup, it's imperative to first back up the 'source' resource upon which others depend, such as the master database, followed by the 'dependent' resources, like the slave database.
see the proposal about orderedResouces in velero: https://github.com/vmware-tanzu/velero/blob/c9e1ade1f7c3b52ac3a38ae853e79a507b33bf95/design/Implemented/backup-resources-order.md

File: pkg/apis/backups/v1alpha1/util.go (hunk 11 above)
Comment (Xieql) [reply pkg/apis/backups/v1alpha1/util.go:41]: yes.

Alternatively, we could allow both to coexist and then take the union.

File: pkg/apis/backups/v1alpha1/restore_type.go (hunk 9 above)
Comment (Xieql) [reply pkg/apis/backups/v1alpha1/restore_type.go:72]: ok

File: pkg/apis/backups/v1alpha1/restore_type.go (hunk 10 above)
Comment (Xieql) [reply pkg/apis/backups/v1alpha1/restore_type.go:81]: ok

File: pkg/apis/backups/v1alpha1/migrate_type.go (hunk 6 above)
Comment (Xieql) [reply pkg/apis/backups/v1alpha1/migrate_type.go:54]: in source cluster

I will fix it

File: pkg/apis/backups/v1alpha1/backup_type.go (hunk 2 above)
Comment (Xieql) [reply pkg/apis/backups/v1alpha1/backup_type.go:42]: Okay, so we actually have a constraint. For the backup operation of the fleet, we can only use the object storage on its associated backup plugin storage.

File: pkg/apis/backups/v1alpha1/backup_type.go (hunk 1 above)
Comment (Xieql) [reply pkg/apis/backups/v1alpha1/backup_type.go:78]: ok

File: pkg/apis/backups/v1alpha1/util.go (hunk 11 above)
Comment (Xieql) [reply pkg/apis/backups/v1alpha1/util.go:41]: We can simplify the design by removing the `ClusterSelector`. If both `Fleet` and `Clusters` are specified, we will take their union as the final set of clusters.

**Review by hzxuzhonghu — APPROVED** [review]
Please rebase

**Comment by Xieql** [comment]
> Please rebase

don'edone

**Pull request merged**